<a href="https://colab.research.google.com/github/Rajat-Sri/Python-Coding-Practice/blob/main/Python_Coding_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Basic Operations In Python



![](https://i.imgur.com/gvSnw4A.png)



## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter.


In [1]:
2+3+9

14

In [2]:
99-73

26

In [3]:
34*-56.7

-1927.8000000000002

In [4]:
3018/14

215.57142857142858

In [5]:
100//7

14

In [6]:
5**3

125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


## Variables:
 While working with a programming language such as Python, information is stored in variables. You can think of variables as containers for storing data. The data stored within a variable is called its value.

In [8]:
a = 4
b = 5
c = a+b
c

9

 **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines.

Inline and single-line comments start with `#`, whereas multi-line comments begin and end with three quotes, i.e. `"""`. Here are some examples of code comments:

In [9]:
my_favorite_number = 1 # an inline comment

In [10]:
# This comment gets its own line
my_least_favorite_number = 3

In [11]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5

## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.

In [12]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [13]:
# Equality check - True
my_favorite_number == 1

True

In [14]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [15]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [16]:
# Not equal check - False
a_neutral_number != 3

False

In [17]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [18]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [19]:
# Less than check - False
my_least_favorite_number < my_favorite_number

False

In [20]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [21]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

False

Just like arithmetic operations, the result of a comparison operation can also be stored in a variable.

In [22]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |

In [23]:
my_favorite_number

1

In [24]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [25]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [26]:
True and False

False

In [27]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [28]:
a_neutral_number = 3

In [29]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [30]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [31]:
my_favorite_number < 0 or True

True

The `not` operator returns `False` if a condition is `True` and `True` if the condition is `False`.

In [32]:
not a_neutral_number == 3

False

In [33]:
not my_favorite_number < 0

True

Logical operators can be combined to form complex conditions. Use round brackets or parentheses `(` and `)` to indicate the order in which logical operators should be applied.

In [34]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [35]:
not (True and 0 < 1) or (False and True)

False

If parentheses are not used, logical operators are applied from left to right.

In [36]:
not True and 0 < 1 or False and True

False

# Variables and Data Types

A Quick Tour of Variables and Data Types in Python


![](https://i.imgur.com/6cg2E9Q.png)


Storing information using variables

In [37]:
my_favorite_color = "blue"

In [38]:
my_favorite_color

'blue'

You can also assign values to multiple variables in a single statement by separating the variable names and values with commas.

In [39]:
color1, color2, color3 = "red", "green", "blue"

In [40]:
color1

'red'

In [41]:
color3

'blue'

You can assign the same value to multiple variables by chaining multiple assignment operations within a single statement.

In [42]:
color4 = color5 = color6 = "magenta"

In [43]:
color5

'magenta'

In [44]:
color6

'magenta'

You can change the value stored within a variable by assigning a new value to it using another assignment statement. Be careful while reassigning variables: when you assign a new value to the variable, the old value is lost and no longer accessible.

In [45]:
my_favorite_color

'blue'

In [46]:
my_favorite_color = "red"

In [47]:
my_favorite_color

'red'

While reassigning a variable, you can also use the variable's previous value to compute the new value.

In [68]:
counter = 10

In [69]:
counter = counter + 1

In [70]:
counter

11

Python provides a shorthand syntax for it.

In [71]:
# Same as `counter = counter + 4`
counter += 4

In [72]:
counter

15

Variable names can be short (`a`, `x`, `y`, etc.) or descriptive ( `my_favorite_color`, `profit_margin`, `the_3_musketeers`, etc.). However, you must follow these rules while naming Python variables:

* A variable's name must start with a letter or the underscore character `_`. It cannot begin with a number.
* A variable name can only contain lowercase (small) or uppercase (capital) letters, digits, or underscores (`a`-`z`, `A`-`Z`, `0`-`9`, and `_`).
* Variable names are case-sensitive, i.e., `a_variable`, `A_Variable`, and `A_VARIABLE` are all different variables.

Here are some valid variable names:

In [73]:
a_variable = 23
is_today_Saturday = False
my_favorite_car = "Delorean"
the_3_musketeers = ["Athos", "Porthos", "Aramis"]

Let's try creating some variables with invalid names. Python prints a syntax error if your variable's name is invalid

In [74]:
a variable = 23

SyntaxError: ignored

In [75]:
is_today_$aturday = False

SyntaxError: ignored

In [76]:
my-favorite-car = "Delorean"

SyntaxError: ignored

In [77]:
3_musketeers = ["Athos", "Porthos", "Aramis"]

SyntaxError: ignored

## Built-in data types in Python

Any data or information stored within a Python variable has a *type*. You can view the type of data stored within a variable using the `type` function.

In [78]:
a_variable

23

In [79]:
type(a_variable)

int

In [80]:
is_today_Saturday

False

In [81]:
type(is_today_Saturday)

bool

In [82]:
the_3_musketeers

['Athos', 'Porthos', 'Aramis']

In [83]:
type(the_3_musketeers)

list

Booleans are automatically converted to `int`s when used in arithmetic operations. `True` is converted to `1` and `False` is converted to `0`.

In [84]:
5 + False

5

In [86]:
3.0 + True

4.0

Any value in Python can be converted to a Boolean using the `bool` function.

Only the following values evaluate to `False` (they are often called *falsy* values):

1. The value `False` itself
2. The integer `0`
3. The float `0.0`
4. The empty value `None`
5. The empty text `""`
6. The empty list `[]`
7. The empty tuple `()`
8. The empty dictionary `{}`
9. The empty set `set()`
10. The empty range `range(0)`

Everything else evaluates to `True` (a value that evaluates to `True` is often called a *truthy* value).

In [87]:
bool(True), bool(1), bool(2.0), bool("hello"), bool([1,2]), bool((2,3)), bool(range(10))

(True, True, True, True, True, True, True)

In [88]:
bool(False), bool(0), bool(0.0), bool(None), bool(""), bool([]), bool(()), bool({}), bool(set()), bool(range(0))

(False, False, False, False, False, False, False, False, False, False)

### None

The None type includes a single value `None`, used to indicate the absence of a value. `None` has the type `NoneType`. It is often used to declare a variable whose value may be assigned later.

In [89]:
nothing = None

In [90]:
type(nothing)

NoneType

You can use single quotes inside a string written with double quotes, and vice versa.

In [91]:
my_favorite_movie = "One Flew over the Cuckoo's Nest"

In [92]:
my_favorite_movie

"One Flew over the Cuckoo's Nest"

In [93]:
my_favorite_pun = 'Thanks for explaining the word "many" to me, it means a lot.'

In [94]:
my_favorite_pun

'Thanks for explaining the word "many" to me, it means a lot.'

To create multiline strings, use three single quotes ''' or three double quotes """ to begin and end the string. Line breaks are represented using the newline character \n

In [97]:
yet_another_pun = '''Son: "Dad, can you tell me what a solar eclipse is?"
Dad: "No sun."'''

In [98]:
yet_another_pun

'Son: "Dad, can you tell me what a solar eclipse is?" \nDad: "No sun."'

Multiline strings are best displayed using the `print` function.

In [99]:
print(yet_another_pun)

Son: "Dad, can you tell me what a solar eclipse is?" 
Dad: "No sun."


You can check the length of a string using the `len` function.

In [100]:
len(my_favorite_movie)

31

A string can be converted into a list of characters using `list` function.

In [113]:
today = 'saturday'

In [114]:
list(today)

['s', 'a', 't', 'u', 'r', 'd', 'a', 'y']

You can access individual characters within a string using the [] indexing notation.

In [115]:
today[3]

'u'

In [116]:
today[-1]

'y'

You can access a part of a string using by providing a `start:end` range instead of a single index in `[]`.

In [117]:
today[2:4]

'tu'

You can also check whether a string contains a some text using the `in` operator.

In [118]:
'a' in today

True

In [119]:
'z' in today

False

Two or more strings can be joined or concatenated using the + operator. Be careful while concatenating strings, sometimes you may need to add a space character " " between words.

In [120]:
full_name = "Derek O'Brien"

In [110]:
greeting = "Hello"

In [111]:
greeting + full_name

"HelloDerek O'Brien"

In [112]:
greeting + " " + full_name + "!" # additional space

"Hello Derek O'Brien!"

Strings in Python have many built-in *methods* that are used to manipulate them. Let's try out some common string methods.

> **Methods**: Methods are functions associated with data types and are accessed using the `.` notation e.g. `variable_name.method()` or `"a string".method()`. Methods are a powerful technique for associating common operations with values of specific data types.

The `.lower()`, `.upper()` and `.capitalize()` methods are used to change the case of the characters.

In [122]:
today = "Saturday"
today

'Saturday'

In [123]:
today.lower()

'saturday'

In [124]:
"saturday".upper()

'SATURDAY'

In [125]:
"monday".capitalize() # changes first character to uppercase

'Monday'

The `.replace` method replaces a part of the string with another string. It takes the portion to be replaced and the replacement text as *inputs* or *arguments*.

In [126]:
another_day = today.replace("Satur", "Wednes")

In [127]:
another_day

'Wednesday'

Note that `replace` returns a new string, and the original string is not modified.

In [128]:
today

'Saturday'

The `.split` method splits a string into a list of strings at every occurrence of provided character(s).

In [129]:
"Sun,Mon,Tue,Wed,Thu,Fri,Sat".split(",")

['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']

The `.strip` method removes whitespace characters from the beginning and end of a string.

In [130]:
a_long_line = "       This is a long line with some space before, after,     and some space in the middle..    "

In [132]:
a_long_line_stripped = a_long_line.strip()

In [134]:
a_long_line_stripped

'This is a long line with some space before, after,     and some space in the middle..'